File: backend/src/agents/protocol.ts
/**
 * Agent 输出协议（P3 落地：Skill 输出协议化）
 *
 * 借鉴 PenguinHarness agent-evaluation 的"纯协议输出"思想：
 * 让每个 Agent 完成任务后在最终回复里输出结构化 YAML 协议块，
 * 下游（前端 / 评测 / 优化 / 管线）据此稳定消费，不再依赖 LLM 自由文本。
 *
 * 协议契约由 instructions 注入（见 buildProtocolContract），
 * Agent 代码只负责解析与校验，二者解耦——协议字段语义由契约文本定义，不在代码里硬编码。
 */
import { parse as parseYaml } from 'yaml'

export interface AgentProtocol {
  /** ok=成功完成，failed=失败 */
  status: 'ok' | 'failed'
  /** 一句话概括完成情况与关键数字 */
  summary: string
  /** 允许 Agent 附加结构化统计字段，供下游交叉验证 */
  [key: string]: unknown
}

/** 追加到 instructions 末尾的输出协议契约（所有 Agent 类型通用） */
export function buildProtocolContract(): string {
  return [
    '## 输出协议（必须遵守）',
    '',
    '完成任务后，你必须在最后一条回复中输出一个 YAML 代码块作为收尾协议，格式如下：',
    '',
    '```yaml',
    'status: ok        # ok 表示成功完成，failed 表示失败',
    'summary: 一句话概括完成情况和关键数字',
    '```',
    '',
    '要求：',
    '- status 只能填 ok 或 failed',
    '- summary 用一句话概括（例如：已保存 12 个分镜）',
    '- 即使中途工具调用失败，也要输出协议并如实填写 status: failed',
  ].join('\n')
}

/** 从 Agent 最终文本中提取并解析协议块 */
export function parseAgentProtocol(text: string): {
  protocol: AgentProtocol | null
  errors: string[]
} {
  if (!text) return { protocol: null, errors: ['empty text'] }

  // 提取 YAML fence 块
  const fenceRe = /```ya?ml\s*\n?([\s\S]*?)```/i
  const match = text.match(fenceRe)
  if (!match) return { protocol: null, errors: ['no yaml fence found'] }

  let parsed: unknown
  try {
    parsed = parseYaml(match[1])
  } catch (err) {
    return { protocol: null, errors: [`yaml parse error: ${(err as Error).message}`] }
  }

  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    return { protocol: null, errors: ['protocol is not an object'] }
  }

  const obj = parsed as Record<string, unknown>
  const status = obj.status
  const summary = obj.summary

  if (status !== 'ok' && status !== 'failed') {
    return { protocol: null, errors: [`invalid status: ${JSON.stringify(status)}`] }
  }
  if (typeof summary !== 'string' || !summary.trim()) {
    return { protocol: null, errors: ['missing or empty summary'] }
  }

  return {
    protocol: { ...obj, status, summary: summary.trim() } as AgentProtocol,
    errors: [],
  }
}

File: backend-py/app/services/agents/protocol.py
"""Agent 输出协议（移植自 ``backend/src/agents/protocol.ts``，79 行）。

**思路**（借鉴 agent-evaluation 的「纯协议输出」）：让每个 Agent 完成任务后在最终回复里输出
一段结构化 **YAML 收尾块**，下游（前端 / 评测 / 优化 / 管线）据此稳定消费，不再依赖 LLM 自由文本。

**解耦**：协议字段语义由 ``build_protocol_contract()`` 的**契约文本**定义（注入 instructions），
代码只负责解析与校验 —— 不在代码里硬编码字段语义。

⚠️ 错误文案是**英文且逐字**（``empty text`` / ``no yaml fence found`` / …），会被上层透出。
⚠️ ``invalid status: …`` 里区分 **缺键**（``undefined``）与**显式 null**（``null``）——
对齐 JS 的 ``JSON.stringify``（``JSON.stringify(undefined)`` 返回 ``undefined`` 本身）。
"""
from __future__ import annotations

import json
import re
from typing import Any

import yaml

__all__ = ["build_protocol_contract", "parse_agent_protocol"]

#: ``\```ya?ml`` 围栏（紧随的 ``\s*\n?`` 与 TS 同形）
_YAML_FENCE = re.compile(r"```ya?ml\s*\n?([\s\S]*?)```", re.IGNORECASE)

#: 区分「键不存在」与「键存在但值为 null」（JS 的 undefined vs null）
_MISSING = object()


def build_protocol_contract() -> str:
    """追加到 instructions 末尾的输出协议契约（**所有 Agent 类型通用**）。"""
    return "\n".join([
        "## 输出协议（必须遵守）",
        "",
        "完成任务后，你必须在最后一条回复中输出一个 YAML 代码块作为收尾协议，格式如下：",
        "",
        "```yaml",
        "status: ok        # ok 表示成功完成，failed 表示失败",
        "summary: 一句话概括完成情况和关键数字",
        "```",
        "",
        "要求：",
        "- status 只能填 ok 或 failed",
        "- summary 用一句话概括（例如：已保存 12 个分镜）",
        "- 即使中途工具调用失败，也要输出协议并如实填写 status: failed",
    ])


def _stringify_status(value: Any) -> str:
    """``JSON.stringify(status)``：缺键 -> ``undefined``、null -> ``null``、字符串 -> 带引号。

    ⚠️ 用**紧凑分隔符**才是真正对齐 ``JSON.stringify``（它从不输出空格）：
    对字符串/标量虽然无差别，但一旦值是个对象，默认的 ``, ``/``: `` 就会多出空格。
    """
    if value is _MISSING:
        return "undefined"
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def parse_agent_protocol(text: str | None) -> dict[str, Any]:
    """从 Agent 最终文本里提取并解析协议块，返回 ``{protocol, errors}``。

    失败时 ``protocol`` 为 None、``errors`` 说明原因（**不抛异常**）。
    """
    if not text:
        return {"protocol": None, "errors": ["empty text"]}

    match = _YAML_FENCE.search(text)
    if match is None:
        return {"protocol": None, "errors": ["no yaml fence found"]}

    try:
        parsed = yaml.safe_load(match.group(1))
    except Exception as err:  # noqa: BLE001 —— 与 TS 的 catch 等价
        return {"protocol": None, "errors": [f"yaml parse error: {err}"]}

    if not isinstance(parsed, dict):
        return {"protocol": None, "errors": ["protocol is not an object"]}

    status = parsed.get("status", _MISSING)
    summary = parsed.get("summary", _MISSING)

    if status not in ("ok", "failed"):
        return {"protocol": None, "errors": [f"invalid status: {_stringify_status(status)}"]}
    if not isinstance(summary, str) or not summary.strip():
        return {"protocol": None, "errors": ["missing or empty summary"]}

    # `{ ...obj, status, summary }`：保留 Agent 附加的统计字段，status/summary 用校验后的值
    protocol = {**parsed, "status": status, "summary": summary.strip()}
    return {"protocol": protocol, "errors": []}
